Implement in-place interpreter call and arithmetic/comparisons

danlliu opened the pull request:
#### d5dfafee77d61cb05dde5f4663cfef89c39233d4
<pre>
Implement in-place interpreter call and arithmetic/comparisons
<a href="https://bugs.webkit.org/show_bug.cgi?id=259378">https://bugs.webkit.org/show_bug.cgi?id=259378</a>
rdar://112627148

Reviewed by Yusuke Suzuki.

Add functionality for call and remaining numeric comparison and arithmetic opcodes.

* JSTests/wasm/ipint-tests/ipint-setup-stress-lol.js: Added.
* JSTests/wasm/ipint-tests/ipint-stress-lol.js: Added.
(import.as.assert.from.string_appeared_here.async test):
* JSTests/wasm/ipint-tests/ipint-stress-lol.wasm: Added.
* JSTests/wasm/ipint-tests/ipint-test-call-arg.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.i32.const.1094795585.local.0.call.inc.call.inc.return.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.local.0.call.inc.call.inc.return.async test): Deleted.
* JSTests/wasm/ipint-tests/ipint-test-call-fparg.js: Copied from JSTests/wasm/ipint-tests/ipint-test-call-arg.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.f32.result.f32.local.0.f32.const.1.f32.add.return.func.export.string_appeared_here.param.f32.result.f32.local.0.call.inc.call.inc.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-f32-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.eq.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.ne.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.lt.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.gt.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.le.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.ge.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-f32-ops.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.abs.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.neg.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.ceil.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.floor.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.trunc.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.nearest.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.sqrt.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.add.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.sub.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.mul.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.div.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.min.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.max.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.copysign.return.close):
(y.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.add.return.async test): Deleted.
* JSTests/wasm/ipint-tests/ipint-test-f64-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.eq.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.ne.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.lt.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.gt.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.le.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.ge.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-i64-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i64.result.i32.local.0.i64.eqz.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.eq.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ne.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.lt_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.lt_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.gt_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.gt_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.le_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.le_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ge_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ge_u.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-recursion.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.recur.export.string_appeared_here.param.i32.result.i32.local.0.i32.eqz.result.i32.then.local.0.else.local.0.i32.const.1.i32.sub.call.recur.local.0.i32.add.return.async test):
* JSTests/wasm/ipint-tests/perf.py:
(progressBar):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32AndLength):
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32): Deleted.
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32AndLengthOfOtherInt32): Deleted.
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::getLocal):
(JSC::Wasm::IPIntGenerator::setLocal):
(JSC::Wasm::IPIntGenerator::addCall):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmIPIntCall):

Canonical link: <a href="https://commits.webkit.org/266214@main">https://commits.webkit.org/266214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0f9d3550a8a95a443ff461bbadc2154c741b7f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13552 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14977 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12605 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16060 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/13580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/15295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13406 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/14069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/11192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15545 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/11357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/11947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11268 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/12432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/12114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15320 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12530 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12608 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/13580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13286 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11879 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/3488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16201 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13668 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1504 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12450 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/3284 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -614,6 +614,12 @@ inline UGPRPair doWasmCall(Wasm::Instance* instance, unsigned functionIndex)
     WASM_CALL_RETURN(instance, codePtr.taggedPtr(), WasmEntryPtrTag);
 }
 
+extern "C" UGPRPair doWasmIPIntCall(Wasm::Instance* instance, unsigned functionIndex);
```
Let's declare it in WasmSlowPaths.h. And ` REFERENCED_FROM_ASM WTF_INTERNAL` annotations are required to avoid removing this function from LTO.

Constellation reviewed (approved):
r=me

danlliu committed:
Implement in-place interpreter call and arithmetic/comparisons
https://bugs.webkit.org/show_bug.cgi?id=259378
rdar://112627148

Reviewed by Yusuke Suzuki.

Add functionality for call and remaining numeric comparison and arithmetic opcodes.

* JSTests/wasm/ipint-tests/ipint-setup-stress-lol.js: Added.
* JSTests/wasm/ipint-tests/ipint-stress-lol.js: Added.
(import.as.assert.from.string_appeared_here.async test):
* JSTests/wasm/ipint-tests/ipint-stress-lol.wasm: Added.
* JSTests/wasm/ipint-tests/ipint-test-call-arg.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.i32.const.1094795585.local.0.call.inc.call.inc.return.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.local.0.call.inc.call.inc.return.async test): Deleted.
* JSTests/wasm/ipint-tests/ipint-test-call-fparg.js: Copied from JSTests/wasm/ipint-tests/ipint-test-call-arg.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.inc.export.string_appeared_here.param.f32.result.f32.local.0.f32.const.1.f32.add.return.func.export.string_appeared_here.param.f32.result.f32.local.0.call.inc.call.inc.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-f32-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.eq.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.ne.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.lt.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.gt.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.le.return.func.export.string_appeared_here.param.f32.f32.result.i32.local.0.local.1.f32.ge.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-f32-ops.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.abs.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.neg.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.ceil.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.floor.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.trunc.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.nearest.return.func.export.string_appeared_here.param.f32.result.f32.local.0.f32.sqrt.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.add.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.sub.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.mul.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.div.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.min.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.max.return.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.copysign.return.close):
(y.async test):
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.result.f32.local.0.local.1.f32.add.return.async test): Deleted.
* JSTests/wasm/ipint-tests/ipint-test-f64-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.eq.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.ne.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.lt.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.gt.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.le.return.func.export.string_appeared_here.param.f64.f64.result.i32.local.0.local.1.f64.ge.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-i64-comparisons.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.i64.result.i32.local.0.i64.eqz.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.eq.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ne.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.lt_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.lt_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.gt_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.gt_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.le_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.le_u.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ge_s.return.func.export.string_appeared_here.param.i64.i64.result.i32.local.0.local.1.i64.ge_u.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-recursion.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.recur.export.string_appeared_here.param.i32.result.i32.local.0.i32.eqz.result.i32.then.local.0.else.local.0.i32.const.1.i32.sub.call.recur.local.0.i32.add.return.async test):
* JSTests/wasm/ipint-tests/perf.py:
(progressBar):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32AndLength):
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32): Deleted.
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt32AndLengthOfOtherInt32): Deleted.
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::getLocal):
(JSC::Wasm::IPIntGenerator::setLocal):
(JSC::Wasm::IPIntGenerator::addCall):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmIPIntCall):

Canonical link: https://commits.webkit.org/266214@main
